Applications/VLMEval/ContentView.swift

```diff
@@ -0,0 +1,383 @@
+            .padding()
+        #endif
+        .toolbar {
+            ToolbarItem {
+                Label(
+                    "Memory Usage: \(deviceStat.gpuUsage.activeMemory.formatted(.byteCount(style: .memory)))",
+                    systemImage: "info.circle.fill"
+                )
+                .labelStyle(.titleAndIcon)
+                .padding(.horizontal)
+                .help(
+                    Text(
+                        """
+                        Active Memory: \(deviceStat.gpuUsage.activeMemory.formatted(.byteCount(style: .memory)))/\(GPU.memoryLimit.formatted(.byteCount(style: .memory)))
+                        Cache Memory: \(deviceStat.gpuUsage.cacheMemory.formatted(.byteCount(style: .memory)))/\(GPU.cacheLimit.formatted(.byteCount(style: .memory)))
+                        Peak Memory: \(deviceStat.gpuUsage.peakMemory.formatted(.byteCount(style: .memory)))
+                        """
+                    )
+                )
+            }
+            ToolbarItem(placement: .primaryAction) {
+                Button {
+                    Task {
+                        copyToClipboard(llm.output)
+                    }
+                } label: {
+                    Label("Copy Output", systemImage: "doc.on.doc.fill")
+                }
+                .disabled(llm.output == "")
+                .labelStyle(.titleAndIcon)
+            }
+        }
+        .task {
+            self.prompt = llm.modelConfiguration.defaultPrompt
+            _ = try? await llm.load()
+        }
+    }
+
+    private func generate() {
+        Task {
+            if let selectedImage = selectedImage {
+                #if os(iOS)
+                    let ciImage = CIImage(image: selectedImage)
+                    await llm.generate(prompt: prompt, image: ciImage ?? CIImage())
+                #else
+                    if let cgImage = selectedImage.cgImage(
+                        forProposedRect: nil, context: nil, hints: nil)
+                    {
+                        let ciImage = CIImage(cgImage: cgImage)
+                        await llm.generate(prompt: prompt, image: ciImage)
+                    }
+                #endif
+            } else if let imageURL = currentImageURL,
+                let ciImage = await loadImageAsCIImage(from: imageURL)
+            {
+                await llm.generate(prompt: prompt, image: ciImage)
+            }
+        }
+    }
+
+    #if os(macOS)
+        private func loadImage(from url: URL) throws -> Data {
+            guard url.startAccessingSecurityScopedResource() else {
+                throw NSError(
+                    domain: "FileAccess", code: -1,
+                    userInfo: [NSLocalizedDescriptionKey: "Failed to access the file."])
+            }
+            defer { url.stopAccessingSecurityScopedResource() }
+            return try Data(contentsOf: url)
+        }
+    #endif
+
+    private func loadImageAsCIImage(from url: URL) async -> CIImage? {
```

Comment:
Why not use:

- https://developer.apple.com/documentation/coreimage/ciimage/1437908-init

```swift
CIIimage(contensOf: URL)
```

Reply:
I end up with the following runtime warning: 

```
Synchronous URL loading of https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg should not occur on this application's main thread as it may lead to UI unresponsiveness. Please switch to an asynchronous networking API such as URLSession.
```

Reply:
ah, interesting!